### PR TITLE
Add exploit for CVE-2019-4716, unauth rce for IBM PA / TM1

### PR DESCRIPTION
**Pull request by pedrib**
This module exploits a vulnerability in IBM TM1 / Planning Analytics that allows an unauthenticated attacker to perform a configuration overwrite.
It starts by quering the Admin server for the available applications, picks one, and then exploits it. You can also provide an application name to bypass this.
The configuration overwrite is used to change an application server authentication method to "CAM", a proprietary IBM auth method, which is simulated by the exploit.
The exploit then performs a fake authentication as admin, and finally abuses TM1 scripting to perform a command injection as root or SYSTEM.

### CODE REVIEW

File: modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+        and then exploits it. You can also provide an application name to bypass this step,
+        and exploit the application directly.
+        The configuration overwrite is used to change an application server authentication
+        method to "CAM", a proprietary IBM auth method, which is simulated by the exploit.
+        The exploit then performs a fake authentication as admin, and finally abuses TM1
+        scripting to perform a command injection as root or SYSTEM.
+        Testing was done on IBM PA 2.0.6 and IBM TM1 10.2.2 on Windows and Linux.
+        Versions up to and including PA 2.0.8 are vulnerable. It is likely that versions
+        earlier than TM1 10.2.2 are also vulnerable (10.2.2 was released in 2014).
+        The RPOR
```
Comment (bcoles) [inline modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb]:
```suggestion
```

**Comment by pedrib** [comment]
@bcoles all fixed. Do you want pcaps? Unless you have access to a paid copy of the software you won't be able to test.

**Comment by bcoles** [comment]
> 
> 
> @bcoles all fixed. Do you want pcaps? Unless you have access to a paid copy of the software you won't be able to test.

pcaps will likely be required and can be sent to `msfdev@metasploit.com`.

I gave the module a quick once over. It is likely that someone else will also wish to give it a once over before landing, especially as the module contains some frowned-upon-but-not-explicitly-forbidden practices like overwriting the `datastore` options.

File: modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+                # Vulnerability discovery and Metasploit module
+          'Gareth Batchelor <gbatchelor@cloudtrace.com.au>'
+                # Real world exploit testing and feedback
+        ],
+      'References'     =>
+        [
+          [ 'CVE', '2019-4716' ],
+          [ 'URL', 'https://www.ibm.com/support/pages/node/1127781' ],
+          [ 'URL', 'https://raw.githubusercontent.com/pedrib/PoC/master/advisories/ibm-tm1-rce.txt' ],
+          [ 'URL', 'FULL_DISC' ]
```
Comment (bcoles) [inline modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb]: Placeholder comment as a reminder for placeholder URL pending publishing.

File: modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+      'Targets'        =>
+        [
+          [ '(Windows) IBM TM1 <= 10.2.2 / Planning Analytics <= 2.0.8',
+            { 'Platform'  => 'win' }
+          ],
+          [ '(Linux) IBM TM1 <= 10.2.2 / Planning Analytics <= 2.0.8',
+            { 'Platform'  => 'linux' }
+          ],
+        ],
+        'Stance'        => Msf::Exploit::Stance::Aggressive,
```
Comment (bcoles) [inline modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb]:
```suggestion
      'Stance'        => Msf::Exploit::Stance::Aggressive,
```

**Comment by pedrib** [comment]
> > @bcoles all fixed. Do you want pcaps? Unless you have access to a paid copy of the software you won't be able to test.
> 
> pcaps will likely be required and can be sent to `msfdev@metasploit.com`.
> 
> I gave the module a quick once over. It is likely that someone else will also wish to give it a once over before landing, especially as the module contains some frowned-upon-but-not-explicitly-forbidden practices like overwriting the `datastore` options.

Yes, this is a common problem for me. It's marked as a TODO in the HttpServer file, mostly because of me:
https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/http/server.rb#L143

This is the third time I had to use a workaround, here are the other two:
https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/linux/http/dlink_hnap_login_bof.rb#L256
https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/windows/nuuo/nuuo_cms_sqli.rb#L122

File: modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb (hunk 2 above)
Comment (bcoles) [inline modules/exploits/multi/misc/ibm_tm1_unauth_rce.rb]:
```suggestion
          [ 'URL', 'https://seclists.org/fulldisclosure/2020/Mar/44' ]
```

**Comment by pedrib** [comment]
@bcoles added ARCH_CMD to it, tested and it's working perfectly. Probably very useful for AV evasion. One thing though - I don't see a payload for Linux where I can set my own command such as cmd/windows/generic ?

**Comment by bcoles** [comment]
> 
> 
> @bcoles added ARCH_CMD to it, tested and it's working perfectly. Probably very useful for AV evasion. One thing though - I don't see a payload for Linux where I can set my own command such as cmd/windows/generic ?

`ARCH_CMD` payloads for Linux platforms are defined as `'platform' => 'unix'`. For generic command execution with a `CMD` option, the payload you're looking for is [cmd/unix/generic](https://github.com/rapid7/metasploit-framework/blob/master/modules/payloads/singles/cmd/unix/generic.rb).

**Comment by pedrib** [comment]
got it, hold on a bit, making some changes

**Comment by bcoles** [comment]
> 
> 
> got it, hold on a bit, making some changes


Usually multiple targets are defined, allowing one `target` for *nix command execution, one for Windows command execution, one for Windows dropper stager, one for Linux dropper stager, etc.

Here's an example for a cross-platform `multi` exploit:

```
      'Targets'         =>
        [
          ['Unix (In-Memory)',
            'Platform'       => 'unix',
            'Arch'           => ARCH_CMD,
            'DefaultOptions' => {'PAYLOAD' => 'cmd/unix/reverse'},
            'Type'           => :unix_memory
          ],
          ['Linux (Dropper)',
            'Platform'       => 'linux',
            'Arch'           => [ARCH_X86, ARCH_X64],
            'DefaultOptions' => {'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp'},
            'Type'           => :linux_dropper
          ],
          ['PowerShell (In-Memory)',
            'Platform'       => 'win',
            'Arch'           => [ARCH_X86, ARCH_X64],
            'DefaultOptions' => {'PAYLOAD' => 'windows/meterpreter/reverse_tcp'},
            'Type'           => :psh_memory
          ],
          ['Windows (In-Memory)',
            'Platform'       => 'win',
            'Arch'           => ARCH_CMD,
            'DefaultOptions' => {'PAYLOAD' => 'cmd/windows/reverse_powershell'},
            'Type'           => :win_memory
          ],
          ['Windows (Dropper)', 
            'Platform'       => 'win',
            'Arch'           => [ARCH_X86, ARCH_X64],
            'DefaultOptions' => {'PAYLOAD' => 'windows/meterpreter/reverse_tcp'},
            'Type'           => :win_dropper
          ]
        ],
```

**Comment by pedrib** [comment]
Is there a problem if I put them all on two targets and let the user choose a payload? Would make it less confusing IMO

**Comment by pedrib** [comment]
```
      'Arch'           =>
        [
          ARCH_CMD,
          ARCH_X86,
          ARCH_X64
        ],
      'Targets'        =>
        [
          [ '(Windows) IBM TM1 <= 10.2.2 / Planning Analytics <= 2.0.8',
            { 'Platform'  => 'win' }
          ],
          [ '(Linux) IBM TM1 <= 10.2.2 / Planning Analytics <= 2.0.8',
            { 'Platform'  => %w{linux unix} }
          ],
        ],
```

**Comment by pedrib** [comment]
Just thought about one thing - it is likely this exploit also works on AIX targets, but I don't have one to test

**Comment by bcoles** [comment]
> 
> 
> Is there a problem if I put them all on two targets and let the user choose a payload? Would make it less confusing IMO

That's up to whoever reviews this PR.

The module will need to be updated to perform dynamic detection of the user-provided payload type, ie, `payload.arch`, and control execution flow. A payload with arch `cmd` will need to execute `execute_command` whereas a `payload` with arch `X64` will need to execute `execute_cmdstager`.

Generally it is cleaner and less confusing to have one conditional `switch/case` for target selection and let the framework take care of verifying whether the supplied `payload` is compatible with the selected `target`, rather than trying to use custom code.

**Comment by pedrib** [comment]
ok understood, I'll make the changes

**Comment by bcoles** [comment]
Here's an example for handling the `target`.

This example is a little more complex than necessary, as it takes advantage of a vulnerability where the response from the command can be viewed, so has custom code to choose whether the payload should be executed in the background or foreground so the operator can see the output.

```ruby
    case target['Type']
    when :unix_memory
      if datastore['PAYLOAD'] == 'cmd/unix/generic'
        execute_command(payload.encoded, foreground: true)
      else 
        execute_command(payload.encoded)
      end
    when :win_memory
      if datastore['PAYLOAD'] == 'cmd/windows/generic'
        execute_command(payload.encoded, foreground: true)
      else
        execute_command(payload.encoded)
      end
    when :psh_memory
      execute_command(
        cmd_psh_payload(
          payload.encoded,
          payload_instance.arch.first,
          { :remove_comspec => true, :encode_final_payload => true }
        )
      )
    when :linux_dropper
      execute_cmdstager(:linemax => 1_500)
    when :win_dropper
      execute_cmdstager(:linemax => 1_500)
    end     
```

**Comment by pedrib** [comment]
@bcoles all done! Let me know your thoughts. BTW I tried sending pcaps to the email address you gave above, but they keep bouncing back. I guess the provider is Gmail, which doesn't like pcaps

**Comment by pedrib** [comment]
I'm using payload.raw for the ARCH_CMD handling, should I use payload.encoded? This doesn't make sense, since the command is executed directly, so it should be raw?

**Comment by bcoles** [comment]
> 
> 
> I'm using payload.raw for the ARCH_CMD handling, should I use payload.encoded? This doesn't make sense, since the command is executed directly, so it should be raw?

Payload encoding is best practice and preferred where possible. Only very rarely is `payload.raw` used (in very restricted circumstances). Using `payload.raw` will prevent the operator from selecting custom payload encodings.

**Comment by bcoles** [comment]
> 
> 
> @bcoles all done! Let me know your thoughts. BTW I tried sending pcaps to the email address you gave above, but they keep bouncing back. I guess the provider is Gmail, which doesn't like pcaps

Strange. That's the correct address. I don't have access to the mailbox, but I known PCAP attachments have worked fine in the past. You could try renaming to `.txt` and/or compressing it.

**Comment by pedrib** [comment]
> > I'm using payload.raw for the ARCH_CMD handling, should I use payload.encoded? This doesn't make sense, since the command is executed directly, so it should be raw?
> 
> Payload encoding is best practice and preferred where possible. Only very rarely is `payload.raw` used (in very restricted circumstances). Using `payload.raw` will prevent the operator from selecting custom payload encodings.

Actually encoding doesn't make any sense here (it's a straight up command injection), but I'll change to it payload.encoded anyway.

**Comment by bcoles** [comment]
> 
> 
> Just thought about one thing - it is likely this exploit also works on AIX targets, but I don't have one to test

Fortunately, the `target` for unix `ARCH_CMD` should also (probably) work on AIX.

You could easily add another `target` for AIX platform. Usually the fact that it's been untested should be documented somewhere. Either in the `target` name, ie `AIX (Untested)`, or in a code comment somewhere nearby.

**Comment by pedrib** [comment]
ok, split the linux and aix command targets.

I have tried sending it encrypted inside a zip, with encrypted file names, and I get this back:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of
its recipients. This is a permanent error. The following address(es)
failed:

msfdev@metasploit.com:
SMTP error from remote server for TEXT command, host: aspmx.l.google.com  reason: 552-5.7.0 This message was blocked because its content presents a potential
552-5.7.0 security issue. Please visit
552-5.7.0  https://support.google.com/mail/?p=BlockedMessage to review our
552 5.7.0 message content and attachment content guidelines. 
 - gsmtp

**Comment by bcoles** [comment]
>
>
> I have tried sending it encrypted inside a zip, with encrypted file names, and I get this back:
> This message was created automatically by mail delivery software.

welp. whoever reviews this PR may have a preferred delivery method. Perhaps a private link via slack or dropbox. If the contents of the pcap aren't sensitive, you could drop it here.

> 
> 
> > > I'm using payload.raw for the ARCH_CMD handling, should I use payload.encoded? This doesn't make sense, since the command is executed directly, so it should be raw?
> > 
> > 
> > Payload encoding is best practice and preferred where possible. Only very rarely is `payload.raw` used (in very restricted circumstances). Using `payload.raw` will prevent the operator from selecting custom payload encodings.
> 
> Actually encoding doesn't make any sense here (it's a straight up command injection), but I'll change to it payload.encoded anyway.

The inverse is true. Using `payload.encoded` is best practice and a point of pride, as this indicates you've effectively enumerated all bad characters and have a stable working exploit, allowing the operator to choose any encoding they wish.

Comparatively, using `payload.raw` is a warning sign that the code has not been thoroughly written or tested, and may indicate that you were defeated by input filtering or validation (or were too lazy to bypass it).

```
$ grep -rn payload.raw modules/ | wc -l
53
$ grep -rn payload.encoded modules/ | wc -l
1793
```

**Comment by pedrib** [comment]
Most likely a properly configured IPS (with specific rules to catch this) can detect this attack, since there is no way to configure SSL on the authentication (hence why the SSL datastore switch).

But an AV might have a hard time identifying something malicious going on, especially if using the Command targets.

**Comment by pedrib** [comment]
@bcoles inspired by you I decided to hunt for bad chars, and found one in Linux and several in Windows. I've tested them on my bench against both OS and seems to be working fine now.

**Comment by wvu-r7** [comment]
@pedrib: Got your pcap and looked through it. Thanks!

**Comment by wvu-r7** [comment]
Module doc updated as per `tools/dev/msftidy_docs.rb` in https://github.com/rapid7/metasploit-framework/commit/8209a4e6afc623306d71ff6dbecfb565e8550e0e.

**Comment by wvu-r7** [comment]
# Release Notes

This adds an exploit for CVE-2019-4716 against IBM Planning Analytics powered by TM1.

**Comment by pedrib** [comment]
@wvu-r7 thanks!

What about this?

Another thing - it seems that now all commands that I send are processed as if they have bad chars, even if they don't. Is there a way to turn that off? I think that might cause problems for long commands.

Is there a fix for it?

For example, if I enter calc.exe as the command, it still runs it under powershell...

**Comment by wvu-r7** [comment]
@pedrib: Show me? Also, are you "careta" on IRC?

**Pull request merged**